aplib.net-demo/Assets/Scripts/WFC/Grid.cs

```diff
@@ -76,5 +111,119 @@ public void PlaceRoom(int x, int y, Room room)
+        public ICollection<Cell> GetConnectedNeighbours(Cell cell)
+        {
+            ICollection<Cell> connectedNeighbours = new Collection<Cell>();
+            ICollection<Cell> neighbours = Get4NeighbouringCells(cell); // Note: no diagonal neighbours
+            foreach (Cell neighbour in neighbours)
+            {
+                if      (cell.Tile.CanConnectInDirection(1) && neighbour.X > cell.X && neighbour.Tile.CanConnectInDirection(3))
+                    connectedNeighbours.Add(neighbour);
+                else if (cell.Tile.CanConnectInDirection(3) && neighbour.X < cell.X && neighbour.Tile.CanConnectInDirection(1))
+                    connectedNeighbours.Add(neighbour);
+                else if (cell.Tile.CanConnectInDirection(0) && neighbour.Y > cell.Y && neighbour.Tile.CanConnectInDirection(2))
+                    connectedNeighbours.Add(neighbour);
+                else if (cell.Tile.CanConnectInDirection(2) && neighbour.Y < cell.Y && neighbour.Tile.CanConnectInDirection(0))
+                    connectedNeighbours.Add(neighbour);
+            }
+
+            return connectedNeighbours;
+        }
+
+        /// <summary>
+        /// This method goes through all cells in this grid, to determine all connected components.
+        /// </summary>
+        /// <returns>A collection of sets of cells, where the sets of cells represent the connected components.</returns>
+        public IList<ISet<Cell>> DetermineConnectedComponents()
+        {
+            ISet<Cell> unvisitedCells = new HashSet<Cell>(_cells.Where(cell => cell.Tile is not Empty)); // Deep copy
+            IList<ISet<Cell>> connectedComponents = new List<ISet<Cell>>();
+
+            while (unvisitedCells.Any())
+            {
+                ISet<Cell> connectedComponent = new HashSet<Cell>();
+                connectedComponents.Add(connectedComponent);
+
+                // Determine connected component, which updates unvisitedCells and connectedComponent
+                DetermineSingleConnectedComponent(unvisitedCells, connectedComponent, unvisitedCells.First());
+            }
+
+            return connectedComponents;
+        }
+
+        /// <summary>
+        /// Given a cell, this method determines the connected component to which the cell belongs.
+        /// </summary>
+        /// <param name="searchSpace">The cells which are considered to be possibly connected to <paramref name="cell"/>.</param>
+        /// <param name="connectedComponent">The <see cref="ISet{Cell}"/> in which the connected component is stored.</param>
+        /// <param name="cell">The cell to start searching from.</param>
+        /// <remarks>The connected component is stored in <paramref name="connectedComponent"/>.</remarks>
+        public void DetermineSingleConnectedComponent(in ISet<Cell> searchSpace, in ISet<Cell> connectedComponent, Cell cell)
+        {
+            ISet<Cell> unvisitedCells = searchSpace;
```

Comment:
This line is unnecessary as it is a proxy to searchSpace. Searchspace should be used throughout the function.

Reply:
That's what I had at first, but the caller then has to specify `unvisited cells` and he might think "well, why not all of them?" Whilst the xml doc might explain this, a proper name is better than hoping someone reads the comments.

Reply:
Isn't it logical semantically to name the parameter `searchSpace`? The search space would naturally shrink whenever we visit a node, as we won't have to search it anymore.

Reply:
+1